### PR TITLE
Add APU DSP Routine

### PR DESCRIPTION
**Pull request by melefo**
The echo functions, the voices, theirs envelope and the BRR and the update are done which allows us to create the sample buffer that allows us to play music.

### CODE REVIEW

File: sources/APU/DSP/DSP.hpp [hunk 1]
```diff
@@ -9,61 +9,108 @@
...
-		uint8_t flg;
+	class APU;
+	struct MemoryMap;
+}
 
-		//! @brief Echo feedback register
-		uint8_t efb;
+namespace ComSquare::APU::DSP
+{
+	enum Envelope : uint {
```
Comment (zoriya) [inline sources/APU/DSP/DSP.hpp:21]: Missing documentation.

File: sources/APU/DSP/DSP.hpp [hunk 2]
```diff
@@ -9,61 +9,108 @@
...
+		Attack,
+		Decay,
+		Sustain
+	};
 
+	struct Master {
+		//! @brief Main Volume register (MVOL)
+		std::array<uint8_t, 2> volume;
+		//! @brief Mutes all channel (6th bit FLG)
+		bool mute : 1;
```
Comment (zoriya) [inline sources/APU/DSP/DSP.hpp:36]: Why are these bools on a single bit?

File: sources/APU/APU.hpp [hunk 3]
```diff
@@ -138,12 +137,19 @@ namespace ComSquare::APU
...
+		//! @brief Renderer used to play sounds
+		Renderer::IRenderer &_renderer;
 
 		//! @brief Internal APU memory separated according to their utility
 		std::shared_ptr<MemoryMap> _map;
 
+		//! @brief Total size of the buffer containing samples
+		static constexpr int32_t bufferSize = 0x10000;
+		//! @brief Buffer containing samples to be played
+		int16_t _soundBuffer[bufferSize];
```
Comment (zoriya) [inline sources/APU/APU.hpp]: A std::array should be used here instead of a raw pointer.

File: sources/APU/DSP/DSP.cpp [hunk 4]
```diff
@@ -257,350 +256,545 @@ namespace ComSquare::APU::DSP
...
+			this->_echo.FIR[7] = data;
 			break;
 		default:
 			throw InvalidAddress("DSP Registers write", addr);
 		}
 	}
+	uint8_t DSP::_readRAM(uint24_t addr) {
+		switch (addr) {
+			case 0x0000 ... 0x00EF:
+				return this->_map.lock()->Page0.read(addr);
```
Comment (zoriya) [inline sources/APU/DSP/DSP.cpp:577]: If _map has expired, this will dereference nullptr and segfault.

**Review by Octopus773 — APPROVED** [review]

File: sources/APU/DSP/DSP.cpp (hunk 4 above)
Comment (zoriya) [reply sources/APU/DSP/DSP.cpp:577]: This should throw a runtime exception & not an InvalidAddress (InvalidAdress will open the debugger if throwned)

File: sources/Renderer/IRenderer.hpp [hunk 5]
```diff
@@ -30,6 +31,11 @@ namespace ComSquare
 			//! @param snes The snes game object (to call the update method).
 			//! @param maxFPS The number of FPS you aim to run on.
 			virtual void createWindow(SNES &snes, int maxFPS) = 0;
+
+			//! @brief Playing all samples from buffer
+			//! @param samples Buffer containing samples
+			//! @param sampleCount number of samples inside buffer
+			virtual void playAudio(std::span<int16_t> samples, uint64_t sampleCount) = 0;
```
Comment (zoriya) [inline sources/Renderer/IRenderer.hpp:38]: The sampleCount is unecesarry since std::span already stores the size.

**Review by zoriya — APPROVED** [review]

**Pull request merged**